### PR TITLE
cdc: support envelope=wrapped json payloads for changefeed queries.

### LINKED ISSUE #717: Support changefeed with query that have envelope='wrapped' option.

### PR DESCRIPTION
**Pull request by sravotto**
Previously, we supported changefeed queries with the default envelope=raw.
 
However, to emit all the values from a deleted row, in addition to the primary keys, the changefeed require the use of the envelope=wrapped, format=json, and diff options. See https://www.cockroachlabs.com/docs/stable/cdc-queries#capture-delete-messages.
    
This change removes support for envelope=raw in favor of envelope=wrapped.
Also, `event_op() AS __event__` is not longer needed, and it must to be removed from the query.

Existing changefeed that use queries must be updated accordingly, using this syntax:

```
CREATE CHANGEFEED
INTO '{scheme}://{host}:{port}/{target_database}/{target_schema}/{target_table}?{query_parameters}'
WITH envelope='wrapped', format='json', diff [additional options]
AS SELECT
  projection
FROM
  table
[WHERE predicate];
 ```
 
Fixes #717.

- - -
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cdc-sink/741)

### CODE REVIEW

**Review by bobvawter — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cdc-sink/741)* status: 1 of 4 files reviewed, 5 unresolved discussions (waiting on @BramGruneir and @sravotto)

___
*[`internal/source/cdc/query_payload.go` line 157 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsDxHPq7RPqiO9tKN_Q:-NsDxHPq7RPqiO9tKN_R:b-k4t3su) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L157)):*
> ```Go
> 		ts, ok := msg.Get(updated)
> 		if !ok {
> 			return errors.New("missing timestamp")
> ```

Is this message going to be user-visible?  It would help the user to diagnose bad payload delivery if this, and the other error messages in this section, read something like `errors.Errorf"could not find timestamp in field %s while attempting to parse envelope=wrapped", updated)`.  The error above should also mention that it's trying to parse an `envelope=raw` payload.
___
*[`internal/source/cdc/query_payload.go` line 164 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsDxwU-Cu_J67sy12Yy:-NsDxwU-Cu_J67sy12Yz:b-d6p32m) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L164)):*
> ```Go
> 		}
> 		var err error
> 		if q.updated, err = hlc.Parse(tsAsString); err != nil {
> ```

`hlc.Time.UnmarshalJSON()` exists, so it might not be necessary to continue to call `json.Unmarshal` here.  Also, is this `hlc.Parse()` and above there's a `decodeUpdatedTimestamp()` function call ?
___
*[`internal/source/cdc/query_payload.go` line 202 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsDyvLvEyjghiMLZ8pL:-NsDyvLvEyjghiMLZ8pM:b-ehyylj) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L202)):*
> ```Go
> 	q.keyValues = make([]json.RawMessage, q.keys.Len())
> 	return q.keys.Range(func(k ident.Ident, pos int) error {
> 		if msg == nil {
> ```

Should this check be a couple lines above?
___
*[`internal/source/cdc/server/integration_test.go` line 68 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsE03_j7QZewZGGffaE:-NsE03_j7QZewZGGffaF:b-qui2p6) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/server/integration_test.go#L68)):*
> ```Go
> 	diff        bool
> 	immediate   bool
> 	rawEnvelope bool
> ```

I'm not seeing where this is set to true to enable this testing; you'd need to update the loop at the top level of `TestIntegration()` that generates the configurations.
___
*[`internal/source/cdc/server/integration_test.go` line 120 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsE0SM6DWHw9ZxZGHDy:-NsE0SM6DWHw9ZxZGHDz:b-875k1z) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/server/integration_test.go#L120)):*
> ```Go
> 		tcs[i] = testConfig{
> 			diff:      i&testModeDiff == testModeDiff,
> 			immediate: i&testModeImmediate == testModeImmediate,
> ```

Here.

**Review by BramGruneir — CHANGES_REQUESTED** [review]
I'm wondering if we should just pick one instead of support both.  @bobvawter , thoughts?

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-:-NsE1YUP2ysmVPW0JL3-:b-gzo5kw)* status: all files reviewed, 13 unresolved discussions (waiting on @sravotto)

___
*[`internal/source/cdc/query_payload.go` line 91 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsDpU0NCFCRKI4XK5Nf:-NsDpU0NCFCRKI4XK5Ng:b-9cz7gl) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L91)):*
> ```Go
> 	// category since we unmarshalled the values below.
> 	var after, before json.RawMessage
> 	if q.after != nil && q.operation != deleteOp {
> ```

Why did you add the after check here?  If after is nil, and the operation isn't delete, what happens?
___
*[`internal/source/cdc/query_payload.go` line 126 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsDpttt3DT_lfzYFis2:-NsDpttuEUq6axOJNztz:b1i1lfo) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L126)):*
> ```Go
> // {"__event__": "insert", "pk" : 42, "v" : 9, "__crdb__": {"updated": "1.0"}}
> // Example wrapped:
> // {"after":{"__event__":"insert","k":2,"v":"a"},"before":null,"updated":"1.0"}
> ```

Can you add a few more examples here?  Ideally for update and delete?  To both raw and wrapped?  And if they have the same values, even better.
___
*[`internal/source/cdc/query_payload.go` line 135 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsDqAPlBsvsetPFv3-T:-NsDqAPlBsvsetPFv3-U:b-8r583p) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L135)):*
> ```Go
> 	}
> 
> 	// Check if there is a __crdb__ property.
> ```

can you rephrase?  It just disjointed and hard to follow.
___
*[`internal/source/cdc/query_payload.go` line 139 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsDqTBv1WkpGyryW65y:-NsDqTBv1WkpGyryW65z:b-fp1tmm) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L139)):*
> ```Go
> 	// And use the top level properties for after.
> 	crdbRaw, hasCrdb := msg.Get(crdbLabel)
> 	if hasCrdb {
> ```

This if statement can be merged with the .Get() function as crdbRaw and hasCrdb are not use outside of the code block.
___
*[`internal/source/cdc/query_payload.go` line 142 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsDyGlKApTerI_nda8P:-NsDyGlKApTerI_nda8Q:bz5aa6z) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L142)):*
> ```Go
> 		var err error
> 		if q.updated, err = decodeUpdatedTimestamp(crdbRaw); err != nil {
> 			return err
> ```

Should this error be errors.Wrapf?
___
*[`internal/source/cdc/query_payload.go` line 165 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsE0EONAP067P941whr:-NsE0EOO4iLBqRB4Wqpu:b-7h39o0) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L165)):*
> ```Go
> 		var err error
> 		if q.updated, err = hlc.Parse(tsAsString); err != nil {
> 			return err
> ```

wrap?
___
*[`internal/source/cdc/query_payload.go` line 179 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsE0R7Z9zzJGbcFsNkd:-NsE0R7Z9zzJGbcFsNke:b-4oi3fs) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L179)):*
> ```Go
> 	}
> 	// if the envelope=wrapped, q.after could be empty
> 	if q.after != nil {
> ```

Can you add a check to ensure that it is only empty if it's wrapped?  

Is there a test for it?  

Also, the comment is about it being empty, so perhaps check for that first instead of not nil?
___
*[`internal/source/cdc/query_payload_test.go` line 38 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsE17wo8NwVSn21vROm:-NsE17wp5neR4Psj0nuK:bpzrch1) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload_test.go#L38)):*
> ```Go
> 		wantErr string
> 	}{
> 		{"insert wrapped",
> ```

Please go through each of these and rename them if they're wrapped or not.  Just for clarity.

**Review by bobvawter — COMMENTED** [review]
Yes, I agree we should just force the issue of using `envelope=wrapped` and return a useful error message for raw payloads.  It's reasonable for use to require a specific format.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-:-NsE6aRT2kFzuzhWD5O6:bdo30kw)* status: all files reviewed, 13 unresolved discussions (waiting on @sravotto)

**Review by sravotto — COMMENTED** [review]
Makes sense. I updated the code to only support `wrapped`. It is a breaking change, and I will update the wiki once the code is merged.

Please take another look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-:-NsKbdNl7Xr_KZ_KcjZt:b-887s9q)* status: 0 of 8 files reviewed, 13 unresolved discussions (waiting on @bobvawter and @BramGruneir)

___
*[`internal/source/cdc/query_payload.go` line 91 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsDpU0NCFCRKI4XK5Nf:-NsKR9f00Euyh9dlahDy:bogwvc1) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L91)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Why did you add the after check here?  If after is nil, and the operation isn't delete, what happens?
</blockquote></details>

I rewrote this based on the fact that we will only support envelope="wrapped". 
If after is nil, then it is a delete.
___
*[`internal/source/cdc/query_payload.go` line 126 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsDpttt3DT_lfzYFis2:-NsJM7jkBZzpzDuqkGBq:b-896fix) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L126)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Can you add a few more examples here?  Ideally for update and delete?  To both raw and wrapped?  And if they have the same values, even better.
</blockquote></details>

Done.
___
*[`internal/source/cdc/query_payload.go` line 135 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsDqAPlBsvsetPFv3-T:-NsKSPNd6fnkqDPgTguk:b-896fix) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L135)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

can you rephrase?  It just disjointed and hard to follow.
</blockquote></details>

Done.
___
*[`internal/source/cdc/query_payload.go` line 139 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsDqTBv1WkpGyryW65y:-NsLVKhe8V_qzYU2k2sn:b-896fix) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L139)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

This if statement can be merged with the .Get() function as crdbRaw and hasCrdb are not use outside of the code block.
</blockquote></details>

Done.
___
*[`internal/source/cdc/query_payload.go` line 142 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsDyGlKApTerI_nda8P:-NsKSZpb4xccsSj3qGfn:b-fk1166) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L142)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Should this error be errors.Wrapf?
</blockquote></details>

no longer applicable.
___
*[`internal/source/cdc/query_payload.go` line 157 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsDxHPq7RPqiO9tKN_Q:-NsJNTxm64QdOqmLPdMs:b-896fix) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L157)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Is this message going to be user-visible?  It would help the user to diagnose bad payload delivery if this, and the other error messages in this section, read something like `errors.Errorf"could not find timestamp in field %s while attempting to parse envelope=wrapped", updated)`.  The error above should also mention that it's trying to parse an `envelope=raw` payload.
</blockquote></details>

Done.
___
*[`internal/source/cdc/query_payload.go` line 164 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsDxwU-Cu_J67sy12Yy:-NsKSA470cVn0EkmyrE9:b-896fix) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L164)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

`hlc.Time.UnmarshalJSON()` exists, so it might not be necessary to continue to call `json.Unmarshal` here.  Also, is this `hlc.Parse()` and above there's a `decodeUpdatedTimestamp()` function call ?
</blockquote></details>

Done.
___
*[`internal/source/cdc/query_payload.go` line 165 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsE0EONAP067P941whr:-NsKSqnIB6rtOmc_BDB7:b-896fix) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L165)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

wrap?
</blockquote></details>

Done.
___
*[`internal/source/cdc/query_payload.go` line 179 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsE0R7Z9zzJGbcFsNkd:-NsKZyha9E9OWQOJWrlD:b7j5u8h) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L179)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Can you add a check to ensure that it is only empty if it's wrapped?  

Is there a test for it?  

Also, the comment is about it being empty, so perhaps check for that first instead of not nil?
</blockquote></details>

Only supporting wrapped now.
___
*[`internal/source/cdc/query_payload.go` line 202 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsDyvLvEyjghiMLZ8pL:-NsKbagHDFWEWxLFmIIu:b-896fix) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload.go#L202)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Should this check be a couple lines above?
</blockquote></details>

Done.
___
*[`internal/source/cdc/query_payload_test.go` line 38 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsE17wo8NwVSn21vROm:-NsK_G386V47TgHwJbTI:bes5iy9) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/query_payload_test.go#L38)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Please go through each of these and rename them if they're wrapped or not.  Just for clarity.
</blockquote></details>

only supporting wrapped now.
___
*[`internal/source/cdc/server/integration_test.go` line 68 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsE03_j7QZewZGGffaE:-NsK_NuY7fTsAGFNUpEP:b-fk1166) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/server/integration_test.go#L68)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

I'm not seeing where this is set to true to enable this testing; you'd need to update the loop at the top level of `TestIntegration()` that generates the configurations.
</blockquote></details>

no longer applicable.
___
*[`internal/source/cdc/server/integration_test.go` line 120 at r1](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsE0SM6DWHw9ZxZGHDy:-NsK_SKt2YewSPM1oLPb:b-fk1166) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/7575c44fac206ed7845fe40ed9043fe5682f96de/internal/source/cdc/server/integration_test.go#L120)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Here.
</blockquote></details>

no longer applicable.

**Review by bobvawter — APPROVED** [review]
:lgtm: w/ a couple of questions.

> Reviewed 3 of 8 files at r2, 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-:-NsPJBe-6bwQK4L98-ex:b4qc9r6)* status: all files reviewed, 12 unresolved discussions (waiting on @BramGruneir and @sravotto)

___
*[`-- commits` line 10 at r3](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsPGa1916bHl9DFGucL):*
:+1: 
___
*[`internal/source/cdc/ndjson.go` line 90 at r3](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsPGiSxD1_THvhPwM27:-NsPGiSxD1_THvhPwM28:b-cget16) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/54794ba57cbc45a9cad0557f0b48861225111073/internal/source/cdc/ndjson.go#L90)):*
> ```Go
> 			return err
> 		}
> 		// Discard phantom deletes.
> ```

Elaborate on why phantom deletes are seen. Is there a CRDB issue or doc to link to?
___
*[`internal/source/cdc/query_payload_test.go` line 97 at r3](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsPIuGS6fGi75htBymE:-NsPIuGS6fGi75htBymF:bu04ct7) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/54794ba57cbc45a9cad0557f0b48861225111073/internal/source/cdc/query_payload_test.go#L97)):*
> ```Go
> 		},
> 		{"phantom delete",
> 			`{"before":null,"updated":"1.0"}`,
> ```

Is is possible to get a message that's just `{"updated":"1.0"}`?
___
*[`internal/source/cdc/webhook_query.go` line 48 at r3](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsPI5sb0y_GxGq-MDda:-NsPI5sb0y_GxGq-MDdb:b-lbvqqy) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/54794ba57cbc45a9cad0557f0b48861225111073/internal/source/cdc/webhook_query.go#L48)):*
> ```Go
> 		Length  int               `json:"length"`
> 		// With envelope="raw" (default for queries), there is a  `__crdb__` property.
> 		Raw []json.RawMessage `json:"__crdb__"`
> ```

Why is this a slice and not just a raw message that you check the length of to see if it's defined?

**Review by BramGruneir — APPROVED** [review]
:lgtm:

> Reviewed 3 of 8 files at r2, 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-:-NsP_O82CNQbPjz6w9jt:bnfp4nl)* status: all files reviewed, 4 unresolved discussions (waiting on @sravotto)

**Review by sravotto — COMMENTED** [review]
I realized that I was using "raw" instead of "bare" for the default envelope. I fixed it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-:-NsPldJD3uY6-8yLCAXC:b-jfj873)* status: 5 of 8 files reviewed, 4 unresolved discussions (waiting on @bobvawter and @BramGruneir)

___
*[`-- commits` line 10 at r3](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsPGa1916bHl9DFGucL):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

:+1:
</blockquote></details>

Done.
___
*[`internal/source/cdc/ndjson.go` line 90 at r3](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsPGiSxD1_THvhPwM27:-NsPlIBs0KFLdWBnj_d5:b-7vv6fd) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/54794ba57cbc45a9cad0557f0b48861225111073/internal/source/cdc/ndjson.go#L90)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Elaborate on why phantom deletes are seen. Is there a CRDB issue or doc to link to?
</blockquote></details>

I created an issue (as an optimization for CRDB). Linked the issue in the query_payload.go
___
*[`internal/source/cdc/query_payload_test.go` line 97 at r3](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsPIuGS6fGi75htBymE:-NsPhDuvBlLhjHjW9Nxy:b-30ztl9) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/54794ba57cbc45a9cad0557f0b48861225111073/internal/source/cdc/query_payload_test.go#L97)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Is is possible to get a message that's just `{"updated":"1.0"}`?
</blockquote></details>

I don't think so, but I have added a test case. Done.
___
*[`internal/source/cdc/webhook_query.go` line 48 at r3](https://reviewable.io/reviews/cockroachdb/cdc-sink/741#-NsPI5sb0y_GxGq-MDda:-NsPhNA6EZjo_WoHywIO:bb6un1j) ([raw file](https://github.com/cockroachdb/cdc-sink/blob/54794ba57cbc45a9cad0557f0b48861225111073/internal/source/cdc/webhook_query.go#L48)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Why is this a slice and not just a raw message that you check the length of to see if it's defined?
</blockquote></details>

Done
